lib/pbench/profiling.conf.py

```diff
@@ -0,0 +1,72 @@
+"""Optional Gunicorn configuration file to enable profiling
+
+This module provides definitions for the pre-request and post-request Gunicorn
+hooks, which allow us to get control before a request begins executing and
+after it completes.  We use these to enable and disable profiling (so that the
+profile reflects only the single method call, with a minimum of extraneous
+execution), and, when the request is complete, we dump the profiling report to
+a file or to the log.
+
+This file is specified via a command line switch on the Gunicorn invocation
+when profiling is requested, and it is read by Gunicorn to set the
+`pre_request` and `post_request` Gunicorn settings to the hook routines
+defined below.
+"""
+
+import cProfile
+from io import StringIO
+import os
+import pstats
+import time
+
+from gunicorn.http import Request
+from gunicorn.workers import base
+
+
+def pre_request(worker: base.Worker, req: Request):
+    """Gunicorn hook function called before a request is executed
+
+    We create a profile context, notify the log that we're profiling (so that
+    no one is surprised), note the current time, and enable Python profiling.
+    (We hang the profile context and time stamp off the existing Gunicorn
+    worker object (ain't Python grand!?), and we use the worker's log.)
+    """
+    worker.profile = cProfile.Profile()
+    worker.log.info(f"PROFILING {worker.pid}: {req.uri}")
```

Comment:
I'm generally bothered by the "single API" profiling, since it relies on "stopping the ride" at just the right time. A more useful alternative might be to provide a way to specify a full or base URI to profile, e.g., `/api/v1/upload/.*` and only profile on a match.